File: 1_C/pset1/credit.c
#include<stdio.h>
//#include<cs50.h>
#include<stdbool.h>

int count(unsigned long long int n);

void card(int n, int initial);

int first_digits(unsigned long long int n);

bool luhn(unsigned long long int n, int digits);

void testcard(unsigned long long int n);

int main(void)
{
    // unsigned long int credit = get_long("Number: ");

    // unsigned long long int credit = 5555555555554444; //MASTERCARD

    unsigned long long int credit = 369421438430814; //INVALIDO

    testcard(credit);
}

int count(unsigned long long int n)
{
    int count = 0;

    while (n != 0)
    {
        n /= 10;

        count++;
    }
    return count;

}

void card(int n, int initial)
{

    if (n == 15 && (initial == 34 || initial == 37))
    {
        printf("AMEX\n");
    }
    else if ((n == 16 || n == 13) && initial / 10 == 4)
    {
        printf("VISA\n");
    }
    else if (n == 16 && (initial == 52 || initial == 53 || initial == 54 || initial == 55 || initial == 51))
    {
        printf("MASTERCARD\n");
    }
    else
    {
        printf("INVALID\n");
    }

}

int first_digits(unsigned long long int n)
{
    unsigned long long int first = n;

    int f;

    while (first >= 100)
    {
        first /= 10;
    } 

    f = first;

    return f;
}


bool luhn(unsigned long long int n, int digits)
{
    int odd = 0;
    int even = 0;

    int last;

    for (int i = 1 ; i < digits + 1 ; i++)
    {
            
        last = n % 10;

        if (i % 2 == 0)
        {

        
            last = 2 * last;



            if (last % 10 != 0 || last == 10)
            {  
                even += (last % 10) + (last / 10); //se tiver 2 digitos o numero deve ser separado
            } 	
            else
            {
                even += last;
            }

        }
        else
        {
            odd += last;
        }
        n /= 10;

    }



    even += odd;


    if (even % 10  == 0)
    {
        return 1;
    }
    else
    {
        return 0;
    }

}

int mypow(int a, int b)
{
    if (b == 1)
    {
        return a;
    }
    else
    {
        return a * mypow(a, b - 1);
    }
}

void testcard(unsigned long long int n)
{
    int tcount = count(n);

    if (luhn(n, tcount) == true)
    {
        card(tcount, first_digits(n));
    }
    else
    {
        printf("INVALID\n");
    }

}

File: 6_Python/pset6/credit.py
import cs50


def main():
    # card = 378282246310005

    # card = 4003600000000014

    card = cs50.get_int("Number: ")
    testcard(card)


def tipo_cartao(n):
    if (int(str(n)[:2]) == 34 or int(str(n)[:2]) == 37) and count(n) == 15:
        print("AMEX")
    elif (int(str(n)[:2]) == 51 or int(str(n)[:2]) == 52 or int(str(n)[:2]) == 53 or int(str(n)[:2]) == 54 or int(str(n)[:2]) == 55) and count(n) == 16:
        print("MASTERCARD")
    elif (int(str(n)[:1]) == 4) and (count(n) == 13 or count(n) == 16):
        print("VISA")
    else:
        print("INVALID")


def cheksumcard(card):
    soma = 0

    temp = card

    # Sum even number from right to left and mutiply by 2
    for i in range(count(card)):

        if i % 2 == 1:

            if count((card % 10) * 2) == 2:

                n = int((card % 10) * 2)
                soma += int(n % 10)
                n = int(n / 10)
                soma += int(n % 10)

            else:

                soma += int((card % 10) * 2)

        card = int(card / 10)

    soma2 = 0

    # Sum odd number from right to left
    for i in range(count(temp)):

        if i % 2 == 0:

            soma2 += temp % 10

        temp = int(temp / 10)

    return soma + soma2


def count(card):
    count = 0
    while (card > 0):
        card = card//10
        count += 1
    return count


def testcard(card):  # Function that test card
    if (cheksumcard(card) % 10) == 0:
        tipo_cartao(card)

    else:
        print("INVALID")


main()
